pkg/splunk/enterprise/util.go

```diff
@@ -34,20 +37,248 @@ import (
+	if version != secretName {
+		// Includes, version extracted, check if version number is valid
+		versionInt, err := strconv.Atoi(version)
+		if err != nil {
+			return -1, errors.New(nonIntegerVersionError)
+		}
+
+		// Versions should be > 0
+		if versionInt <= 0 {
+			return -1, errors.New(lessThanOrEqualToZeroVersionError)
+		}
+
+		return versionInt, nil
+	}
+
+	// Secret name not matching required criteria
+	return -1, fmt.Errorf(nonMatchingStringError, secretName, versionedSecretIdentifier)
+}
+
+// GetExistingLatestVersionedSecret retreives latest EXISTING versionedSecretIdentifier based secret existing currently in the namespace
+func GetExistingLatestVersionedSecret(c splcommon.ControllerClient, namespace string, versionedSecretIdentifier string) (*corev1.Secret, int, bool) {
+	scopedLog := log.WithName("GetExistingLatestVersionedSecret").WithValues(
+		"versionedSecretIdentifier", versionedSecretIdentifier,
+		"namespace", namespace)
+
+	// Get list of secrets in K8S cluster
+	secretList := corev1.SecretList{}
+
+	// Retreive only secrets only from namespace
+	listOpts := []client.ListOption{
+		client.InNamespace(namespace),
+	}
+
+	err := c.List(context.TODO(), &secretList, listOpts...)
+	if err != nil || len(secretList.Items) == 0 {
+		scopedLog.Error(err, "Secrets not found in namespace")
+		return nil, -1, false
+	}
+
+	// secretFound is true when we find atleast one versionedSecretIdentifier based secret AND we can extract its version successfully
+	var secretFound bool = false
+
+	// existingLatestVersion holds the version number of the latest versionedSecretIdentifier based secret, if atleast one exists(defaults to -1)
+	var existingLatestVersion int = -1
+
+	// existingLatestVersionedSecret holds the latest versionedSecretIdentifier based secret, if atleast one exists
+	var existingLatestVersionedSecret corev1.Secret
+
+	// Loop through all secrets in K8S cluster
+	for _, secret := range secretList.Items {
+		// Check if the secret is based on the versionedSecretIdentifier and extract version
+		version, err := GetVersionedSecretVersion(secret.GetName(), versionedSecretIdentifier)
+		if err != nil {
+			// Secret name not matching required criteria, move onto next one
+			continue
+		}
+
+		// Version extracted successfully, checking for latest version
+		if version > existingLatestVersion {
+			// Updating latest version
+			existingLatestVersion = version
+			existingLatestVersionedSecret = secret
+		}
+		secretFound = true
+	}
+
+	return &existingLatestVersionedSecret, existingLatestVersion, secretFound
+}
+
+// GetLatestVersionedSecret is used to create/retreive latest versionedSecretIdentifier based secret, cr is optional for owner references(pass nil if not required)
+func GetLatestVersionedSecret(c splcommon.ControllerClient, cr splcommon.MetaObject, namespace string, versionedSecretIdentifier string) (*corev1.Secret, error) {
+	var latestVersionedSecret *corev1.Secret
+	var err error
+	var secretFound bool
```

Comment:
similar as above, i wonder if we need this extra bool?

Reply:
Done. Removed the bool.